[InstCombine] Fold X * (2^N + 1) >> N -> X + X >> N, or directly to X if X >> N is 0

AZero13 opened the pull request:
Alive2 Proofs:
https://alive2.llvm.org/ce/z/eSinJY
https://alive2.llvm.org/ce/z/sweDgc
https://alive2.llvm.org/ce/z/-2dXZi
https://alive2.llvm.org/ce/z/kx2PhF
https://alive2.llvm.org/ce/z/e9QjM6

topperc commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1411,13 +1411,23 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+        if (BitWidth > 2 && ShAmtC * 2 == BitWidth)
+          return BinaryOperator::CreateAnd(X, ConstantInt::get(Ty, *MulC - 2));
+
+        // lshr (mul (X, 2^N + 1)), N -> add (X, lshr(X, N))
```
Does this transform require `nuw`? This is nested in a `match(Op0, m_NUWMul`.

topperc commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1411,13 +1411,23 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+
+        // lshr (mul (X, 2^N + 1)), N -> add (X, lshr(X, N))
+        if (Op0->hasOneUse()) {
+          auto *NewAdd = BinaryOperator::CreateNUWAdd(
```
You create an NUW add here, but your proof shows an NSW add

topperc commented:
The `tgt` function in your Alive2 proof has a dead instruction in it that is distracting to reviewing the proof.

AtariDreams commented:
> A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> 
> 
> 
> Proof: https://alive2.llvm.org/ce/z/u3y_-F

I addressed this and edited the commit. Please see.

AtariDreams commented:
Also see: https://alive2.llvm.org/ce/z/NGZx9f

topperc commented:
> A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.

Proposed where?

AtariDreams commented:
The ashr is a leftover from a previous PR but they can work as a negative test for now because I am working on another patch that deals with ashr and I might merge it with this one if it is best.

AtariDreams commented:
@topperc I fixed the proofs.

AtariDreams commented:
@dtcxzyw Thoughts on this?

topperc commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1411,13 +1411,24 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+        if (Op0->hasOneUse()) {
+          auto *NewAdd = BinaryOperator::CreateNUWAdd(
+              X, Builder.CreateLShr(X, ConstantInt::get(Ty, ShAmtC), "",
+                                    I.isExact()));
```
This preserves the exact flag, but the exact flag is not used in your proofs

AtariDreams replied:
Now I do.

AtariDreams commented:
@dtcxzyw You think you could run benchmarks on this?

AtariDreams commented:
@dtcxzyw would you be so kind as to benchmark this

AZero13 committed:
[InstCombine] Pre-commit tests (NFC)

AtariDreams commented:
@dtcxzyw would you be so kind as to benchmark this

AtariDreams commented:
@topperc what do you think about his?

AZero13 committed:
[InstCombine] Fold X * (2^N + 1) >> N -> X + X >> N, or directly to X if X >> N is 0

Alive2 Proofs:
https://alive2.llvm.org/ce/z/eSinJY
https://alive2.llvm.org/ce/z/sweDgc
https://alive2.llvm.org/ce/z/-2dXZi
https://alive2.llvm.org/ce/z/kx2PhF
https://alive2.llvm.org/ce/z/e9QjM6

topperc commented:
@AtariDreams why do you keep force pushing patches? The github UI doesn't make it easy to see if you changed anything in the code in the patch or just rebased.

AtariDreams commented:
> @AtariDreams why do you keep force pushing patches? The github UI doesn't make it easy to see if you changed anything in the code in the patch or just rebased. If you don't write any comment to explain the what you did, the reviewer is left guessing.

Edited the commit message is what I did.

nikic commented:
@AtariDreams As you have already been told before: The commit messages do not matter. Only the pull request description will be used. Please refrain from force-pushing to change the commit description.

AtariDreams commented:
Alright, I just fixed a couple of whitespace issues, and now this is good to go!

topperc commented:
> Alright, I just fixed a couple of whitespace issues, and now this is good to go!

And yet again you force pushed despite being repeatedly told not to.

AtariDreams commented:
@dtcxzyw would you be so kind as to benchmark this

AtariDreams commented:
Added negative tests!

AtariDreams commented:
I had to add those negative tests to precommit

AtariDreams commented:
@topperc @dtcxzyw What do you think about this PR?

topperc commented on llvm/test/Transforms/InstCombine/ashr-lshr.ll:
```diff
@@ -604,3 +604,66 @@ define <2 x i8> @ashr_known_pos_exact_vec(<2 x i8> %x, <2 x i8> %y) {
+;
+  %mul = mul nuw i32 %x, 3
+  %res = lshr i32 %mul, 1
+  call void @use (i32 %mul)
```
Drop the space after `@use`

topperc commented:
> > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> 
> Proposed where?

This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.

nikic commented:
> > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> > 
> > 
> > Proposed where?
> 
> This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.

A variant of this PR was originally part of https://github.com/llvm/llvm-project/pull/88193. What was "proposed" is that these unrelated changes be removed from that PR.

AtariDreams commented:
> > > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> 
> > > 
> 
> > > 
> 
> > > Proposed where?
> 
> > 
> 
> > This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.
> 
> 
> 
> A variant of this PR was originally part of https://github.com/llvm/llvm-project/pull/88193. What was "proposed" is that these unrelated changes be removed from that PR.

And those unrelated changes were!

topperc commented:
> > > > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > Proposed where?
> > 
> > 
> > > 
> > 
> > 
> > > This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.
> > 
> > 
> > A variant of this PR was originally part of #88193. What was "proposed" is that these unrelated changes be removed from that PR.
> 
> And those unrelated changes were removed!

Great. But that doesn't change the fact that the description of this PR needs to be modified. It will become a commit message that will live in the LLVM repo forever. It doesn't look like the fact that this code was previously part of a different PR is relevant to understanding this PR so the sentence should just be deleted.

AtariDreams commented:
> > > > > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > Proposed where?
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.
> > > 
> > > 
> > > A variant of this PR was originally part of #88193. What was "proposed" is that these unrelated changes be removed from that PR.
> > 
> > 
> > And those unrelated changes were removed!
> 
> Great. But that doesn't change the fact that the description of this PR needs to be modified. It will become a commit message that will live in the LLVM repo forever. It doesn't look like the fact that this code was previously part of a different PR is relevant to understanding this PR so the sentence should just be deleted.



> > > > > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > Proposed where?
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.
> > > 
> > > 
> > > A variant of this PR was originally part of #88193. What was "proposed" is that these unrelated changes be removed from that PR.
> > 
> > 
> > And those unrelated changes were removed!
> 
> Great. But that doesn't change the fact that the description of this PR needs to be modified. It will become a commit message that will live in the LLVM repo forever. It doesn't look like the fact that this code was previously part of a different PR is relevant to understanding this PR so the sentence should just be deleted.



> > > > > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > Proposed where?
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.
> > > 
> > > 
> > > A variant of this PR was originally part of #88193. What was "proposed" is that these unrelated changes be removed from that PR.
> > 
> > 
> > And those unrelated changes were removed!
> 
> Great. But that doesn't change the fact that the description of this PR needs to be modified. It will become a commit message that will live in the LLVM repo forever. It doesn't look like the fact that this code was previously part of a different PR is relevant to understanding this PR so the sentence should just be deleted.

Fixed!

AtariDreams commented:
> > > > > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > Proposed where?
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.
> > > 
> > > 
> > > A variant of this PR was originally part of #88193. What was "proposed" is that these unrelated changes be removed from that PR.
> > 
> > 
> > And those unrelated changes were removed!
> 
> Great. But that doesn't change the fact that the description of this PR needs to be modified. It will become a commit message that will live in the LLVM repo forever. It doesn't look like the fact that this code was previously part of a different PR is relevant to understanding this PR so the sentence should just be deleted.

I removed the sentence from the commit!

AtariDreams commented:
@topperc I believe this is ready!

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1457,13 +1457,24 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+        if (ShAmtC * 2 == BitWidth)
+          return BinaryOperator::CreateAnd(X, ConstantInt::get(Ty, *MulC - 2));
+
+        // lshr (mul nuw (X, 2^N + 1)), N -> add nuw (X, lshr(X, N))
```
As this transform introduces multiple uses of X, do we need a freeze here?
I don't know why alive2 doesn't complain about this.

nikic replied:
I'm not entirely sure, but it's possible that the nuw flag makes this safe (by converting problematic undef cases to poison).

dtcxzyw replied:
cc @regehr @nunoplopes 

Can you guys have a look? Thank you!

dtcxzyw replied:
> I'm not entirely sure, but it's possible that the nuw flag makes this safe (by converting problematic undef cases to poison).

It is confusing to me. Is this behavior documented in LangRef?
See https://alive2.llvm.org/ce/z/i2eXpk
```
; bad
define i8 @src1(i8 %x) {
  %shl = shl i8 %x, 1
  ret i8 %shl
}

define i8 @tgt1(i8 %x) {
  %add = add i8 %x, %x
  ret i8 %add
}

; good
define i8 @src2(i8 %x) {
  %shl = shl nuw i8 %x, 1
  ret i8 %shl
}

define i8 @tgt2(i8 %x) {
  %add = add i8 %x, %x
  ret i8 %add
}
```

regehr replied:
Nuno will have to enlighten us, I don't see why `nuw` would stop multiple uses of undef from being a problem, that doesn't make any sense to me

nunoplopes replied:
`shl nuw %x, 1` implies that %x is not undef. This is because if %x is undef, we can pick a value that makes the shift yield poison.
Then we can refine poison with `add %x, %x`.

TL;DR:  `shl nuw undef, 1 -> poison -> add %x, %x`

dtcxzyw replied:
> `shl nuw %x, 1` implies that %x is not undef. This is because if %x is undef, we can pick a value that makes the shift yield poison. Then we can refine poison with `add %x, %x`.
> 
> TL;DR: `shl nuw undef, 1 -> poison -> add %x, %x`

It makes sense to me. Thank you!

dtcxzyw reviewed (approved):
LGTM if https://github.com/llvm/llvm-project/pull/90295/files#r1602406923 is not an issue. Thank you!

nikic commented on llvm/test/Transforms/InstCombine/ashr-lshr.ll:
```diff
@@ -604,3 +604,66 @@ define <2 x i8> @ashr_known_pos_exact_vec(<2 x i8> %x, <2 x i8> %y) {
+  ret i32 %lshr
+}
+
+declare void @use(i32)
```
This is missing tests for numbers other than `3` and `1`. You should test both another combination that works and one that doesn't.

It's also worth noting that this can still be further generalized. You now handle something like `* 3 >> 1`, but this also holds for `* 6 >> 2` or `* 12 >> 3`.

AtariDreams replied:
For that one, I was in fact thinking about that, but I think that could best be done in another PR because why stop there?

* 6 >> 2 is just * 3 / 2. 

We could do the same with x * 150 / 100.

AtariDreams replied:
I am in fact working on this in another PR but that is taking more work than expected, so I figure at least do some baby steps so if something has to be reverted, it won't be this PR.

nikic replied:
I agree that doing this in a separate PR makes sense. Just wanted to mention the possibility.

AtariDreams replied:
The point is, I very much considered this, and not just for shifts, as this transform came about when I was reading code that did 50% increases by doing x = x * 150/100

dtcxzyw replied:
> This is missing tests for numbers other than 3 and 1. You should test both another combination that works and one that doesn't.

@AtariDreams You can find more cases here: https://github.com/dtcxzyw/llvm-opt-benchmark/pull/607/files

> It's also worth noting that this can still be further generalized. You now handle something like * 3 >> 1, but this also holds for * 6 >> 2 or * 12 >> 3.

We need a canonicalization that converts `(X *nuw/nsw 6) >> 2 -> (X *nuw/nsw 3) >> 1`.

AtariDreams replied:
> > This is missing tests for numbers other than 3 and 1. You should test both another combination that works and one that doesn't.
> 
> @AtariDreams You can find more cases here: https://github.com/dtcxzyw/llvm-opt-benchmark/pull/607/files
> 
> > It's also worth noting that this can still be further generalized. You now handle something like * 3 >> 1, but this also holds for * 6 >> 2 or * 12 >> 3.
> 
> We need a canonicalization that converts `(X *nuw/nsw 6) >> 2 -> (X *nuw/nsw 3) >> 1`.

I mean, this is just glorified fraction reduction, which I did in fact start writing, but have not been able to get it to work.

AtariDreams replied:
> > > This is missing tests for numbers other than 3 and 1. You should test both another combination that works and one that doesn't.
> > 
> > 
> > @AtariDreams You can find more cases here: https://github.com/dtcxzyw/llvm-opt-benchmark/pull/607/files
> > > It's also worth noting that this can still be further generalized. You now handle something like * 3 >> 1, but this also holds for * 6 >> 2 or * 12 >> 3.
> > 
> > 
> > We need a canonicalization that converts `(X *nuw/nsw 6) >> 2 -> (X *nuw/nsw 3) >> 1`.
> 
> I mean, this is just glorified fraction reduction, which I did in fact start writing, but have not been able to get it to work.

So, I am doing that in multiple steps.

AtariDreams commented:
New tests added!

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1457,13 +1457,24 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+              X, Builder.CreateLShr(X, ConstantInt::get(Ty, ShAmtC), "",
+                                    I.isExact()));
+          if (cast<OverflowingBinaryOperator>(Op0)->hasNoSignedWrap())
+            NewAdd->setHasNoSignedWrap(true);
```
`NewAdd->setHasNoSignedWrap(cast<OverflowingBinaryOperator>(Op0)->hasNoSignedWrap());`

?

AtariDreams replied:
Honestly, I think this looks cleaner, but you do have a point. I should keep the way it was written before.

AtariDreams commented:
@dtcxzyw I do not have merge permissions! But I think this is now ready!

AtariDreams commented:
Is this good?

AtariDreams commented:
@nikic can we merge this?

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1483,6 +1494,16 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+                                  I.isExact()));
+      }
+    }
+
```
At least imo, the code would be simpler as:
```
if (match(Op0, m_NUWMul(m_Value(X), m_APInt(MulC))) && BitWidth > 2 &&
    (*MulC - 1).isPowerOf2() && MulC->logBase2() == ShAmtC) {
  auto *BO0 = cast<OverflowingBinaryOperator>(Op0);
  // Look for a "splat" mul pattern - it replicates bits across each half
  // of a value, so a right shift is just a mask of the low bits:
  // lshr i[2N] (mul nuw X, (2^N)+1), N --> and iN X, (2^N)-1
  if (ShAmtC * 2 == BitWidth && BO0->hasNoUnsignedWrap())
    return BinaryOperator::CreateAnd(X, ConstantInt::get(Ty, *MulC - 2));

  // lshr (mul nuw (X, 2^N + 1)), N -> add nuw (X, lshr(X, N))
  if (Op0->hasOneUse() &&
      (BO0->hasNoUnsignedWrap() || &&BO0->hasNoSignedWrap())) {
    auto *NewAdd = BinaryOperator::CreateAdd(
        X,
        Builder.CreateLShr(X, ConstantInt::get(Ty, ShAmtC), "", I.isExact()));
    NewAdd->setHasNoSignedWrap(BO0->hasNoSignedWrap());
    NewAdd->setHasNoUnsignedWrap(BO0->hasNoSignedWrap());
    return NewAdd;
  }
}
```

AtariDreams replied:
Yeah, but I want to keep it the way it is now in case other new folds come up, so it is easier to extend.

goldsteinn replied:
In general simpler code w/ less duplication will be easier to extend.

AtariDreams replied:
> if (match(Op0, m_NUWMul(m_Value(X), m_APInt(MulC))) && BitWidth > 2 &&
>     (*MulC - 1).isPowerOf2() && MulC->logBase2() == ShAmtC) {
>   auto *BO0 = cast<OverflowingBinaryOperator>(Op0);
>   // Look for a "splat" mul pattern - it replicates bits across each half
>   // of a value, so a right shift is just a mask of the low bits:
>   // lshr i[2N] (mul nuw X, (2^N)+1), N --> and iN X, (2^N)-1
>   if (ShAmtC * 2 == BitWidth && BO0->hasNoUnsignedWrap())
>     return BinaryOperator::CreateAnd(X, ConstantInt::get(Ty, *MulC - 2));
> 
>   // lshr (mul nuw (X, 2^N + 1)), N -> add nuw (X, lshr(X, N))
>   if (Op0->hasOneUse() &&
>       (BO0->hasNoUnsignedWrap() || &&BO0->hasNoSignedWrap())) {
>     auto *NewAdd = BinaryOperator::CreateAdd(
>         X,
>         Builder.CreateLShr(X, ConstantInt::get(Ty, ShAmtC), "", I.isExact()));
>     NewAdd->setHasNoSignedWrap(BO0->hasNoSignedWrap());
>     NewAdd->setHasNoUnsignedWrap(BO0->hasNoSignedWrap());
>     return NewAdd;
>   }
> }

this eliminates the MulC >> ShiftC optimization

goldsteinn replied:
That wasn't exact code, just the idea of putting all the `m_Mul` and `(*MulC - 1).isPowerOf2() && MulC->logBase2() == ShAmtC` together to simplify the code.

AtariDreams commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1483,6 +1494,16 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+            X, Builder.CreateLShr(X, ConstantInt::get(Ty, ShAmtC), "",
+                                  I.isExact()));
+      }
+    }
```
Did an investigation and yes this is true but there are cases where it doesn't have to be exact either and I think that's best for a follow up PR

AtariDreams replied:
Fixed!

AtariDreams commented:
@topperc @dtcxzyw @nikic I addressed all your comments and concerns. Can we please merge this?

AtariDreams commented:
Putting this in hold for a tiny bit as I do one last thing.

AtariDreams commented:
@nikic I moved the fold to X special cases to analysis, and the rest remain intact. This now is good to merge!

AtariDreams commented:
@dtcxzyw What do you think about this?

AtariDreams commented:
@dtcxzyw Can we please benchmark this one more time?

topperc commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -1479,6 +1479,30 @@ static Value *simplifyLShrInst(Value *Op0, Value *Op1, bool IsExact,
+        MulC->logBase2() == ShAmtC) {
+      // FIXME: This condition should be covered by the computeKnownBits, but
+      // for some reason it is not, so keep this in for now. This has no
+      // negative affects, but KnownBits should be able to infer a number of
```
affects -> effects

topperc commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -1479,6 +1479,30 @@ static Value *simplifyLShrInst(Value *Op0, Value *Op1, bool IsExact,
 
+  // Look for a "splat" mul pattern - it replicates bits across each half
+  // of a value, so a right shift is just a mask of the low bits:
+  // lshr i[2N] (mul nuw X, (2^N)+1), N --> and iN X, (2^N)-1
```
This code doesn't create an `and`

topperc commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -1479,6 +1479,30 @@ static Value *simplifyLShrInst(Value *Op0, Value *Op1, bool IsExact,
+        return X;
+      const KnownBits XKnown = computeKnownBits(X, /* Depth */ 0, Q);
+      if (XKnown.countMaxActiveBits() <= ShAmtC)
+        return X;
```
The computeKNownBits part of this doesn't appear to be tested. I cloned your branch and commented this out and nothing failed.

I would think calling computeKnownBits on the lshr would return 0 for this case all by itself.

AtariDreams replied:
it is folding 

 %xx = and i32 %x, 360
  %m = mul i32 %x, 65537
  %t = ashr i32 %m, 16
  
  to ; CHECK-NEXT:    [[M:%.*]] = mul nuw i32 [[X:%.*]], 65537
; CHECK-NEXT:    [[T:%.*]] = ashr i32 [[M]], 16
; CHECK-NEXT:    ret i32 [[T]]

instead of just folding to the and

topperc replied:
This is in `simplifyLShrInst` so why are you showing an `ashr`?

AtariDreams replied:
Fixed it!

topperc replied:
Nevermind. Forget what I said.

nikic reviewed (changes requested):
This PR looked sensible the last time I checked it -- now it contains another new random InstSimplify fold and also seems to change some one-use conditions.

Please learn some patience. Don't keep making random changes to your PR just because nobody is around to hit the merge button right this second. Do not ping PRs more often than once a week.

nikic commented:
You are modifying an existing fold to produce a different result (in a different pass now no less) and adding a new fold in the same PR. That's not going to fly. Please submit two PRs for the two changes.
